### ISSUE TITLE
Consumer hang in PollFetches after metadata leader epoch goes backwards

### ISSUE THREAD

**Issue by jcsp** [twmb/franz-go#198]
## Context

Bit of a long description here because it's something happening deep inside a scale test intermittently that had to be carefully teased out.

This is manifesting as consumers mysteriously disappearing from a consumer group without any error surfacing on the client side (https://github.com/redpanda-data/redpanda/issues/5959).

It may be related to https://github.com/redpanda-data/redpanda/issues/5435, in which we're not sure why the client was seeing epochs go backward.

The test program in use is kgo-repeater from https://github.com/redpanda-data/kgo-verifier, with 3 client machines and each machine running a kgo-repeater process with 128 franz-go clients inside it.

The test in use is ManyPartitionsTest (https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/scale_tests/many_partitions_test.py)

## Failure Behavior

The test restarts all cluster nodes concurrently, then, in sequence:
- waits for them to all become responsive to admin api-
- waits for all partitions to have a leader
- waits for our consumer group to report state Stable with the expected number of members.

The test fails intermittently on the third check, where we see a few fewer members than expected (a typical number would be 378 group members where we expected 384).

Once we figure out exactly which workers have dropped out by comparing the group with the expected set, we can go get the debug logs from the clients, including franz-go debug logs (this is with a hacked version of the test for the extra debug).

The resulting log for one of these zombie consumers is: here [kgo-consumer-hang.log.gz](https://github.com/twmb/franz-go/files/9453156/kgo-consumer-hang.log.gz)

##  Analysis

Filtering the log down to the client code's debug logs and franz-go's RPC read/write lines, and filtering out Produce messages, we get this as the "last gasp" before it stops progressing:

```
time="2022-08-30T12:59:42Z" level=debug msg="Consume ip-172-31-43-24_6564_w_69 fetching..."
ip-172-31-43-24_6564_w_69[DEBUG] wrote Metadata v7; broker: 4, bytes_written: 58, write_wait: 19.745µs, time_to_write: 12.503µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read Metadata v7; broker: 4, bytes_read: 0, read_wait: 99.405µs, time_to_read: 3.1µs, err: EOF
ip-172-31-43-24_6564_w_69[DEBUG] read from broker errored, killing connection; addr: ip-172-31-32-186:9092, broker: 4, successful_reads: 1, err: EOF
ip-172-31-43-24_6564_w_69[DEBUG] wrote Metadata v7; broker: 8, bytes_written: 58, write_wait: 2.839588ms, time_to_write: 12.344µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] wrote OffsetCommit v7; broker: 1, bytes_written: 2621, write_wait: 2.158705ms, time_to_write: 11.184µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read Metadata v7; broker: 8, bytes_read: 599842, read_wait: 19.079µs, time_to_read: 712.058368ms, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read OffsetCommit v7; broker: 1, bytes_read: 220, read_wait: 3.366566ms, time_to_read: 501.237326ms, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] wrote FindCoordinator v2; broker: 9, bytes_written: 50, write_wait: 7.677401ms, time_to_write: 11.076µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read FindCoordinator v2; broker: 9, bytes_read: 42, read_wait: 2.623459ms, time_to_read: 2.253479ms, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] wrote OffsetCommit v7; broker: 1, bytes_written: 2621, write_wait: 14.784µs, time_to_write: 8.487µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read OffsetCommit v7; broker: 1, bytes_read: 220, read_wait: 625.193µs, time_to_read: 123.969535ms, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] wrote FindCoordinator v2; broker: 1, bytes_written: 50, write_wait: 10.737µs, time_to_write: 8.852µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read FindCoordinator v2; broker: 1, bytes_read: 42, read_wait: 14.675µs, time_to_read: 402.18207ms, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] wrote OffsetCommit v7; broker: 1, bytes_written: 2621, write_wait: 11.355µs, time_to_write: 13.713µs, err: <nil>
ip-172-31-43-24_6564_w_69[DEBUG] read OffsetCommit v7; broker: 1, bytes_read: 220, read_wait: 3.594153ms, time_to_read: 294.030554ms, err: <nil>
```

There are some WARN messages for connections refused to brokers, this is an intentional part of the test where the brokers are all restarted.

But after they're all back online, there are some lines like this:
```
ip-172-31-43-24_6564_w_69[DEBUG] metadata leader epoch went backwards, ignoring update; topic: scale_000000, partition: 11864, old_leader_epoch: 2, new_leader_epoch: 2, current_num_rewinds: 1
```

That log itself is confusing because of a small bug where the new epoch is assigned from the old epoch before printing, so old_leader_epoch and new_leader_epoch are always equal.  However it seems like the real bug is that the retry logic isn't happening: current_num_rewinds is 1, the code looks like it's means to retry until that hits 5, but no retrys ever happen: the client just stops here and hangs inside PollFetches.

**Comment by twmb** [comment]
Pushed commit 5f24faef09e40ce653a182cbe0c381737a6d2bbf, that and timestamps can help track down what's up here. Nothing looks outright _wrong_ in the logs yet, but it's difficult to say without timestamps (and with one of the logs being incorrect).

**Comment by jcsp** [comment]
This helpfully reproduced first try with the updated binary, here are the logs from one of the workers that dropped out of a consumer group:
[franz-go-198-take2.log.gz](https://github.com/twmb/franz-go/files/9455699/franz-go-198-take2.log.gz)

For reference the timing of the cluster restart are:
```
[INFO  - 2022-08-30 18:34:20,653 - many_partitions_test - _restart_stress - lineno:455]: Cluster restart 1/10...
[INFO  - 2022-08-30 18:34:29,154 - many_partitions_test - _restart_stress - lineno:458]: Restart 1/10 complete.  Waiting for elections...
[INFO  - 2022-08-30 18:34:44,800 - many_partitions_test - _restart_stress - lineno:466]: Post-restart elections done.
```

**Comment by jcsp** [comment]
Stack output on SIGQUIT from the worker process.  At the time the test failed, there were 128 clients instantiated of which 3 had dropped out of the group
[stack.log](https://github.com/twmb/franz-go/files/9455784/stack.log)

**Comment by twmb** [comment]
This goroutine is problematic:

```
goroutine 1682564 [chan send, 20 minutes]:
github.com/twmb/franz-go/pkg/kgo.(*source).loopFetch(0xc01d6d3950)
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220830181657-5f24faef09e4/pkg/kgo/source.go:517 +0x3e7
created by github.com/twmb/franz-go/pkg/kgo.(*source).maybeConsume
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220830181657-5f24faef09e4/pkg/kgo/source.go:468 +0xdc
```

This is hanging in a spot that should never hang. The end result is that consuming from one broker never finishes, and this blocks the consumer session ending, which blocks consuming entirely.

The commit closing this should fix it, I'll tag 1.7.1 if we can confirm a bit that this avoids the problem.

**Comment by twmb** [comment]
Note that the leader epoch going backwards is unrelated, but the continual movement of partitions makes this issue more likely (every moved partition causes the consumer to reset its work -- stop a session and start a new one).

**Comment by twmb** [comment]
This fix has been released in v1.7.1
